### PR TITLE
Fixes struct padding for TPL #1210

### PR DESCRIPTION
**Pull request by 17zhangw**
Fixes Issue #1210 by inserting extra padding members when constructing structs.

### CODE REVIEW

**Review by pmenon — CHANGES_REQUESTED** [review]
First pass. You also need to add tests.

File: src/execution/ast/context.cpp [hunk 1]
```diff
@@ -277,6 +277,34 @@ MapType *MapType::Get(Type *key_type, Type *value_type) {
   return map_type;
 }
 
+Field StructType::CreatePaddingElement(uint32_t size, Context *ctx, uint32_t *remainder) {
+  // We only need a placeholder name. The identifier for the placeholder
+  // does not have to be unique since TPL code will never refer to a
+  // placeholder field and LLVM IR does not rely on field names.
+  ast::Identifier name = ctx->GetIdentifier("__field$0$");
```
[the remark this answers is not in the document]
Comment (lmwnshn) [reply src/execution/ast/context.cpp:284]: I think the comment is incorrect but the `GetIdentifier` function makes names unique automatically within the context.

[the remark this answers is not in the document]
Comment (pmenon) [reply src/execution/ast/context.cpp:284]: `ast::Context::GetIdentifier()` doesn't unique-ify names.

File: src/execution/ast/type.cpp [hunk 2]
```diff
@@ -105,9 +105,10 @@ MapType::MapType(Type *key_type, Type *val_type)
 // ---------------------------------------------------------
 
 StructType::StructType(Context *ctx, uint32_t size, uint32_t alignment, util::RegionVector<Field> &&fields,
-                       util::RegionVector<uint32_t> &&field_offsets)
+                       util::RegionVector<Field> &&unpadded_fields, util::RegionVector<uint32_t> &&field_offsets)
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply src/execution/ast/type.cpp:108]: Created Issue #1218 to track.

**Review by pmenon — CHANGES_REQUESTED** [review]
Much cleaner! A few small changes, then it should be good.

File: sample_tpl/struct-padding.tpl [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    state.count1 = 42
+}
+fun tearDownState(state: *State) -> nil {
+}
+fun main(execCtx: *ExecutionContext) -> int32 {
+    var state: State
+    state.count2 = 500
+    setUpState(execCtx, &state)
+    tearDownState(&state)
+    return state.count2
```
[the remark this answers is not in the document]
Comment (17zhangw) [reply sample_tpl/struct-padding.tpl:17]: So I tried to simplify this test into primitive types by removing the Sorter instance and using basic types, arrays, and nested structures but they did not return incorrect results when run against the master branch. I noticed that it's only when using certain non-primitives (i.e Sorter or TableVectorIterator for instance) that an incorrect result will be returned.

[the remark this answers is not in the document]
Comment (pmenon) [reply sample_tpl/struct-padding.tpl:17]: I suppose that's good news.

Strip this down further to the bare essentials. For example, `ExecutionContext` isn't needed. Also, add checks in this same file for arrays and nested structures containing `Sorters`.

File: src/execution/ast/context.cpp [hunk 4]
```diff
@@ -30,6 +30,16 @@
 // #include "execution/util/csv_reader.h" Fix later.
 #include "execution/util/execution_common.h"
 
+namespace {
+
+terrier::execution::ast::Field CreatePaddingElement(uint32_t id, uint32_t size, terrier::execution::ast::Context *ctx) {
+  terrier::execution::ast::Identifier name = ctx->GetIdentifier("__pad$" + std::to_string(id) + "$");
+  auto *pad_type = terrier::execution::ast::BuiltinType::Get(ctx, terrier::execution::ast::BuiltinType::Int8);
+  return terrier::execution::ast::Field(name, terrier::execution::ast::ArrayType::Get(size, pad_type));
```
Comment (pmenon) [inline src/execution/ast/context.cpp:38]: Move this into the `terrier::execution::ast` namespace directly above `StructType::Get()` where it's needed. Refactoring this way should clean it up.

File: test/execution/ast_type_test.cpp [hunk 5]
```diff
@@ -19,6 +19,27 @@ class TypeTest : public TplTest {
 
   ast::Identifier Name(const std::string &s) { return Ctx()->GetIdentifier(s); }
 
+  ast::Identifier PaddingName(uint32_t offset) { return Ctx()->GetIdentifier("__pad$" + std::to_string(offset) + "$"); }
+
+  void CheckArrayType(StructType *struct_type, ast::Identifier name, uint32_t offset, uint32_t length,
```
Comment (pmenon) [inline test/execution/ast_type_test.cpp:24]: This interface seems a little unintuitive w.r.t. what it does. Types don't have names, and the name can be derived from the offset, so isn't required in the interface. But, at a higher level, I think what you want is:

```
/** Checks if the element at the given offset is padding of the given size. */
void CheckIsPadding(StructType *type, std::size_t offset, std::size_t expected_padding_size);
```

The implementation pulls out the field by name (using the name derived from the offset) and (1) checks it exists and (2) checks the type is indeed an array of the given size and whose elements are bytes using an another `void CheckIsArrayType(Type *type, std::size_t num_elems, Type *elem_type);`

File: src/execution/ast/context.cpp [hunk 6]
```diff
@@ -300,18 +310,27 @@ StructType *StructType::Get(Context *ctx, util::RegionVector<Field> &&fields) {
...
+        size = new_size;
       }
 
       // Update size and calculate alignment
       field_offsets.push_back(size);
-      size += field.type_->GetSize();
-      alignment = std::max(alignment, field.type_->GetAlignment());
+      all_fields.emplace_back(field);
+      size += field_type->GetSize();
+      alignment = std::max(alignment, field_type->GetAlignment());
```
Comment (pmenon) [inline src/execution/ast/context.cpp:333]: Either use `field_align` or remove that variable.

File: src/execution/ast/context.cpp [hunk 7]
```diff
@@ -300,18 +310,27 @@ StructType *StructType::Get(Context *ctx, util::RegionVector<Field> &&fields) {
...
+    util::RegionVector<Field> all_fields(ctx->GetRegion());
     util::RegionVector<uint32_t> field_offsets(ctx->GetRegion());
+    all_fields.reserve(fields.size());
+    field_offsets.reserve(fields.size());
     for (const auto &field : fields) {
+      auto *field_type = field.type_;
+
       // Check if the type needs to be padded
-      uint32_t field_align = field.type_->GetAlignment();
+      uint32_t field_align = field_type->GetAlignment();
```
Comment (pmenon) [inline src/execution/ast/context.cpp:321]: Nit: `const`-ify?

**Review by pmenon — CHANGES_REQUESTED** [review]
Small refactors.

File: test/execution/ast_type_test.cpp [hunk 8]
```diff
@@ -19,6 +19,25 @@ class TypeTest : public TplTest {
 
   ast::Identifier Name(const std::string &s) { return Ctx()->GetIdentifier(s); }
 
+  void CheckIsArrayType(Type *type, size_t num_elem, Type *elem_type) {
+    EXPECT_EQ(type->GetTypeId(), Type::TypeId::ArrayType);
```
Comment (pmenon) [inline test/execution/ast_type_test.cpp:23]: Change to `EXPECT_TRUE(type->IsArrayType());`

File: src/execution/ast/context.cpp [hunk 9]
```diff
@@ -32,6 +32,12 @@
 
 namespace terrier::execution::ast {
 
+Field CreatePaddingElement(uint32_t id, uint32_t size, Context *ctx) {
```
Comment (pmenon) [inline src/execution/ast/context.cpp:35]: I should've been clearer. Move this into an anonymous namespace **within** `terrier::execution::ast` like [this](https://github.com/cmu-db/noisepage/blob/master/src/execution/ast/type_printer.cpp#L11). Also, move this right above `StructType::Get()`

**Review by pmenon — CHANGES_REQUESTED** [review]
Much better. Last few changes, then it's good to go.

File: sample_tpl/struct-padding.tpl [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    state.count2 = 500
+    state.embed.a = 42
+    return state.count2
+}
+
+fun main() -> int32 {
+    var sum : int32
+    sum = test1() + test2() + test3()
+    return sum
+}
```
Comment (pmenon) [inline sample_tpl/struct-padding.tpl:52]: Can you add some comments on what's going on here?

Comment (17zhangw) [reply sample_tpl/struct-padding.tpl:56]: Added comments.

**Review by pmenon — APPROVED** [review]
LGTM.

**Pull request merged**